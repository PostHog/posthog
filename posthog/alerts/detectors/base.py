"""Base classes and types for alert detectors."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import StrEnum
from typing import Any, Optional


class DetectorType(StrEnum):
    """Available detector types."""

    THRESHOLD = "threshold"
    ZSCORE = "zscore"
    MAD = "mad"  # Median Absolute Deviation


class DetectionDirection(StrEnum):
    """Direction for anomaly detection."""

    UP = "up"  # Alert on values above normal
    DOWN = "down"  # Alert on values below normal
    BOTH = "both"  # Alert on values above or below normal


class ValueType(StrEnum):
    """Whether to run detection on raw values or deltas."""

    RAW = "raw"  # Run detection on raw metric values
    DELTA = "delta"  # Run detection on period-over-period changes


@dataclass
class DetectionResult:
    """Result of running a detector on data."""

    value: Optional[float]  # The raw metric value
    detector_score: Optional[float]  # The detector-specific score (e.g., z-score, mad score)
    is_breach: bool  # Whether this constitutes a breach
    breach_messages: list[str]  # Human-readable breach descriptions
    metadata: dict[str, Any]  # Additional detector-specific info


class BaseDetector(ABC):
    """Base class for all alert detectors."""

    def __init__(self, config: dict[str, Any]):
        """Initialize detector with configuration."""
        self.config = config
        self.validate_config()

    @abstractmethod
    def detect(
        self, values: list[float], series_name: str = "Series", value_type: ValueType = ValueType.RAW
    ) -> DetectionResult:
        """
        Run detection on a list of values.

        Args:
            values: Time series values (most recent last)
            series_name: Name of the series for breach messages
            value_type: Whether to use raw values or deltas

        Returns:
            DetectionResult with breach status and details
        """
        pass

    @abstractmethod
    def validate_config(self) -> None:
        """Validate the detector configuration. Raise ValueError if invalid."""
        pass

    @property
    @abstractmethod
    def detector_type(self) -> DetectorType:
        """Return the type of this detector."""
        pass

    def _compute_deltas(self, values: list[float]) -> list[float]:
        """Compute period-over-period deltas from values."""
        if len(values) < 2:
            return []
        return [values[i] - values[i - 1] for i in range(1, len(values))]
