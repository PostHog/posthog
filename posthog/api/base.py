from rest_framework.pagination import CursorPagination as RestCursorPagination


class CursorPagination(RestCursorPagination):
    ordering = "-created_at"
    page_size = 100
