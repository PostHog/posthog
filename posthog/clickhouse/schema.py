# This file contains all CREATE TABLE queries, used to sync and test schema
import re

from posthog.clickhouse.dead_letter_queue import *
from posthog.clickhouse.plugin_log_entries import *
from posthog.models.cohort.sql import *
from posthog.models.event.sql import *
from posthog.models.group.sql import *
from posthog.models.person.sql import *
from posthog.models.session_recording_event.sql import *

CREATE_MERGETREE_TABLE_QUERIES = (
    CREATE_COHORTPEOPLE_TABLE_SQL,
    PERSON_STATIC_COHORT_TABLE_SQL,
    DEAD_LETTER_QUEUE_TABLE_SQL,
    EVENTS_TABLE_SQL,
    GROUPS_TABLE_SQL,
    PERSONS_TABLE_SQL,
    PERSONS_DISTINCT_ID_TABLE_SQL,
    PERSON_DISTINCT_ID2_TABLE_SQL,
    PLUGIN_LOG_ENTRIES_TABLE_SQL,
    SESSION_RECORDING_EVENTS_TABLE_SQL,
)
CREATE_DISTRIBUTED_TABLE_QUERIES = (
    WRITABLE_EVENTS_TABLE_SQL,
    DISTRIBUTED_EVENTS_TABLE_SQL,
    WRITABLE_SESSION_RECORDING_EVENTS_TABLE_SQL,
    DISTRIBUTED_SESSION_RECORDING_EVENTS_TABLE_SQL,
)
CREATE_KAFKA_TABLE_QUERIES = (
    KAFKA_DEAD_LETTER_QUEUE_TABLE_SQL,
    KAFKA_EVENTS_TABLE_JSON_SQL,
    KAFKA_GROUPS_TABLE_SQL,
    KAFKA_PERSONS_TABLE_SQL,
    KAFKA_PERSONS_DISTINCT_ID_TABLE_SQL,
    KAFKA_PERSON_DISTINCT_ID2_TABLE_SQL,
    KAFKA_PLUGIN_LOG_ENTRIES_TABLE_SQL,
    KAFKA_SESSION_RECORDING_EVENTS_TABLE_SQL,
)
CREATE_MV_TABLE_QUERIES = (
    DEAD_LETTER_QUEUE_TABLE_MV_SQL,
    EVENTS_TABLE_JSON_MV_SQL,
    GROUPS_TABLE_MV_SQL,
    PERSONS_TABLE_MV_SQL,
    PERSONS_DISTINCT_ID_TABLE_MV_SQL,
    PERSON_DISTINCT_ID2_MV_SQL,
    PLUGIN_LOG_ENTRIES_TABLE_MV_SQL,
    SESSION_RECORDING_EVENTS_TABLE_MV_SQL,
)

CREATE_TABLE_QUERIES = (
    CREATE_MERGETREE_TABLE_QUERIES
    + CREATE_DISTRIBUTED_TABLE_QUERIES
    + CREATE_KAFKA_TABLE_QUERIES
    + CREATE_MV_TABLE_QUERIES
)

build_query = lambda query: query if isinstance(query, str) else query()
get_table_name = lambda query: re.findall(r" ([a-z0-9_]+) ON CLUSTER", build_query(query))[0]
