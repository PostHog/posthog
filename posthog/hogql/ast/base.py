import re
from typing import Optional, Literal

from pydantic import BaseModel, Extra

from posthog.hogql.constants import ConstantDataType
from posthog.hogql.errors import NotImplementedException

# NOTE: when you add new AST fields or nodes, add them to the Visitor classes in visitor.py as well!

camel_case_pattern = re.compile(r"(?<!^)(?=[A-Z])")


class AST(BaseModel):
    class Config:
        extra = Extra.forbid

    def accept(self, visitor):
        camel_case_name = camel_case_pattern.sub("_", self.__class__.__name__).lower()
        method_name = f"visit_{camel_case_name}"
        if hasattr(visitor, method_name):
            visit = getattr(visitor, method_name)
            return visit(self)
        if hasattr(visitor, "visit_unknown"):
            return visitor.visit_unknown(self)
        raise NotImplementedException(f"Visitor has no method {method_name}")


class Type(AST):
    def get_child(self, name: str) -> "Type":
        raise NotImplementedException("Type.get_child not overridden")

    def has_child(self, name: str) -> bool:
        return self.get_child(name) is not None

    def resolve_constant_type(self) -> Optional["ConstantType"]:
        from posthog.hogql.ast import UnknownType

        return UnknownType()


class ConstantType(Type):
    data_type: ConstantDataType

    def resolve_constant_type(self) -> "ConstantType":
        return self


class Expr(AST):
    type: Optional[Type]


class Macro(Expr):
    name: str
    expr: Expr
    # Whether the macro is an inlined column "SELECT 1 AS a" or a subquery "SELECT a AS (SELECT 1)"
    macro_format: Literal["column", "subquery"]
