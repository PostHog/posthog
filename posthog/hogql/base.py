import re
from dataclasses import dataclass, field

from typing import Literal, Optional

from posthog.hogql.constants import ConstantDataType
from posthog.hogql.errors import NotImplementedException


# Given a string like "CorrectHorseBS", match the "H" and "B", so that we can convert this to "correct_horse_bs"
camel_case_pattern = re.compile(r"(?<!^)(?<![A-Z])(?=[A-Z])")


@dataclass(kw_only=True)
class AST:
    start: Optional[int] = field(default=None)
    end: Optional[int] = field(default=None)

    def accept(self, visitor):
        camel_case_name = camel_case_pattern.sub("_", self.__class__.__name__).lower()
        if "hog_qlx" in camel_case_name:
            camel_case_name = camel_case_name.replace("hog_qlx", "hogqlx_")
        method_name = f"visit_{camel_case_name}"
        if hasattr(visitor, method_name):
            visit = getattr(visitor, method_name)
            return visit(self)
        if hasattr(visitor, "visit_unknown"):
            return visitor.visit_unknown(self)
        raise NotImplementedException(f"Visitor has no method {method_name}")


@dataclass(kw_only=True)
class Type(AST):
    def get_child(self, name: str) -> "Type":
        raise NotImplementedException("Type.get_child not overridden")

    def has_child(self, name: str) -> bool:
        return self.get_child(name) is not None

    def resolve_constant_type(self) -> Optional["ConstantType"]:
        return UnknownType()


@dataclass(kw_only=True)
class Expr(AST):
    type: Optional[Type] = field(default=None)


@dataclass(kw_only=True)
class CTE(Expr):
    """A common table expression."""

    name: str
    expr: Expr
    # Whether the CTE is an inlined column "WITH 1 AS a" or a subquery "WITH a AS (SELECT 1)"
    cte_type: Literal["column", "subquery"]


@dataclass(kw_only=True)
class ConstantType(Type):
    data_type: ConstantDataType

    def resolve_constant_type(self) -> "ConstantType":
        return self

    def print_type(self) -> str:
        raise NotImplementedException("ConstantType.print_type not implemented")


@dataclass(kw_only=True)
class UnknownType(ConstantType):
    data_type: ConstantDataType = field(default="unknown", init=False)

    def print_type(self) -> str:
        return "Unknown"
