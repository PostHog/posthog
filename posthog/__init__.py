import sys
from unittest.mock import MagicMock
import deltalake.writer

# DLT imports the missing function `try_get_deltatable` from `deltalake` which
# it doesn't actually use but also no longer exists. So we can safely mock
# it to allow the app to startup without module errors
deltalake.writer.try_get_deltatable = MagicMock()  # type: ignore
sys.modules["deltalake.writer"] = deltalake.writer


# This will make sure the app is always imported when
# Django starts so that shared_task will use this app.
from posthog.celery import app as celery_app  # noqa: E402

__all__ = ("celery_app",)

# snowflake-connector-python tries to access a root folder which errors out in pods.
# This sets the snowflake home directory to a relative folder
import os  # noqa: E402

os.environ["SNOWFLAKE_HOME"] = "./.snowflake"
