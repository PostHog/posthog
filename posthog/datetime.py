from datetime import datetime, timedelta


def end_of_day(reference_date: datetime):
    return datetime(
        year=reference_date.year, month=reference_date.month, day=reference_date.day, tzinfo=reference_date.tzinfo
    ) + timedelta(days=1, microseconds=-1)


def start_of_day(reference_date: datetime):
    return datetime(
        year=reference_date.year, month=reference_date.month, day=reference_date.day, tzinfo=reference_date.tzinfo
    )
