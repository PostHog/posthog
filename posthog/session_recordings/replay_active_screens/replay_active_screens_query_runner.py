from datetime import datetime, timedelta
from typing import Optional

from posthog.hogql import ast
from posthog.hogql.parser import parse_select
from posthog.hogql.query import execute_hogql_query
from posthog.hogql_queries.query_runner import AnalyticsQueryRunner
from posthog.schema import (
    CachedReplayActiveScreensQueryResponse,
    ReplayActiveScreensQuery,
    ReplayActiveScreensQueryResponse,
)


class ReplayActiveScreensQueryRunner(AnalyticsQueryRunner[ReplayActiveScreensQueryResponse]):
    query: ReplayActiveScreensQuery
    cached_response: CachedReplayActiveScreensQueryResponse

    def cache_target_age(self, last_refresh: Optional[datetime], lazy: bool = False) -> Optional[datetime]:
        if last_refresh is None:
            return None
        return last_refresh + timedelta(hours=1)

    def to_query(self) -> ast.SelectQuery | ast.SelectSetQuery:
        # Use Python's datetime.now() which respects frozen time in tests
        now = datetime.now()

        query = """
            SELECT
                cutQueryString(cutFragment(url)) as screen,
                count(distinct session_id) as count
            FROM (
                SELECT
                    session_id,
                    arrayJoin(any(all_urls)) as url
                FROM raw_session_replay_events
                WHERE min_first_timestamp >= {python_now} - interval 7 day
                  AND min_first_timestamp <= {python_now}
                GROUP BY session_id
                HAVING date_diff('second', min(min_first_timestamp), max(max_last_timestamp)) > 5
            )
            GROUP BY screen
            ORDER BY count DESC
            LIMIT 10
        """

        with self.timings.measure("parse_select"):
            parsed_select = parse_select(
                query, placeholders={"python_now": ast.Constant(value=now)}, timings=self.timings
            )

        return parsed_select

    def to_actors_query(self) -> ast.SelectQuery | ast.SelectSetQuery:
        return self.to_query()

    def _calculate(self) -> ReplayActiveScreensQueryResponse:
        query = self.to_query()

        response = execute_hogql_query(
            query_type="ReplayActiveScreensQuery",
            query=query,
            team=self.team,
            timings=self.timings,
            modifiers=self.query.modifiers or self.modifiers,
            limit_context=self.limit_context,
            workload=self.workload,
        )

        results = []
        for row in response.results or []:
            results.append(
                {
                    "screen": str(row[0]),
                    "count": row[1],
                }
            )

        return ReplayActiveScreensQueryResponse(
            results=results,
            timings=response.timings,
            types=response.types,
            columns=response.columns,
            hasMore=response.hasMore,
            limit=response.limit,
            offset=response.offset,
            hogql=response.hogql,
        )
