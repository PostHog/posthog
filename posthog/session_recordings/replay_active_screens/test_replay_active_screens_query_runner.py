from datetime import datetime, timedelta
from freezegun import freeze_time
from uuid import uuid4

from posthog.schema import ReplayActiveScreensQuery
from posthog.session_recordings.replay_active_screens.replay_active_screens_query_runner import (
    ReplayActiveScreensQueryRunner,
)
from posthog.session_recordings.queries.test.session_replay_sql import produce_replay_summary
from posthog.test.base import (
    APIBaseTest,
    ClickhouseTestMixin,
    _create_person,
    snapshot_clickhouse_queries,
    flush_persons_and_events,
)


@freeze_time("2024-01-15T12:00:00Z")
class TestReplayActiveScreensQueryRunner(ClickhouseTestMixin, APIBaseTest):
    def setUp(self):
        super().setUp()

    def test_basic_query_returns_active_screens(self):
        # Create some test persons
        _create_person(
            distinct_ids=["user1"],
            team=self.team,
        )
        _create_person(
            distinct_ids=["user2"],
            team=self.team,
        )

        # Create replay events with different URLs within the last 7 days
        base_time = datetime.now()

        # Session 1 with URL https://app.posthog.com/insights
        produce_replay_summary(
            team_id=self.team.pk,
            distinct_id="user1",
            session_id=str(uuid4()),
            first_timestamp=base_time - timedelta(days=1),
            last_timestamp=base_time - timedelta(days=1, seconds=-60),
            all_urls=["https://app.posthog.com/insights", "https://app.posthog.com/insights/123"],
        )

        # Session 2 with URL https://app.posthog.com/dashboard
        produce_replay_summary(
            team_id=self.team.pk,
            distinct_id="user1",
            session_id=str(uuid4()),
            first_timestamp=base_time - timedelta(days=2),
            last_timestamp=base_time - timedelta(days=2, seconds=-60),
            all_urls=["https://app.posthog.com/dashboard"],
        )

        # Session 3 with same URL as session 1 (different user)
        produce_replay_summary(
            team_id=self.team.pk,
            distinct_id="user2",
            session_id=str(uuid4()),
            first_timestamp=base_time - timedelta(days=1),
            last_timestamp=base_time - timedelta(days=1, seconds=-60),
            all_urls=["https://app.posthog.com/insights"],
        )

        # Create a session that's too short (should be filtered out)
        produce_replay_summary(
            team_id=self.team.pk,
            distinct_id="user1",
            session_id=str(uuid4()),
            first_timestamp=base_time - timedelta(days=1),
            last_timestamp=base_time - timedelta(days=1, seconds=-3),
            all_urls=["https://app.posthog.com/features"],
        )

        # Create a session that's too old (should be filtered out)
        produce_replay_summary(
            team_id=self.team.pk,
            distinct_id="user1",
            session_id=str(uuid4()),
            first_timestamp=base_time - timedelta(days=8),
            last_timestamp=base_time - timedelta(days=8, seconds=-60),
            all_urls=["https://app.posthog.com/old"],
        )

        flush_persons_and_events()

        # Run the query
        query = ReplayActiveScreensQuery()
        runner = ReplayActiveScreensQueryRunner(query=query, team=self.team)
        response = runner.calculate()

        # Verify results
        assert len(response.results) > 0

        # Results should be sorted by count descending
        counts = [r.count for r in response.results]
        assert counts == sorted(counts, reverse=True)

        # Find insights screen in results
        insights_result = next((r for r in response.results if "insights" in r.screen), None)
        assert insights_result is not None
        assert insights_result.count == 2  # 2 sessions with insights URLs

        # Find dashboard screen in results
        dashboard_result = next((r for r in response.results if "dashboard" in r.screen), None)
        assert dashboard_result is not None
        assert dashboard_result.count == 1  # 1 session with dashboard URL

    def test_query_filters_short_sessions(self):
        # Create a person
        _create_person(
            distinct_ids=["user_short"],
            team=self.team,
        )

        base_time = datetime.now()

        # Create only short sessions (< 5 seconds)
        produce_replay_summary(
            team_id=self.team.pk,
            distinct_id="user_short",
            session_id="short1",
            first_timestamp=base_time - timedelta(days=1),
            last_timestamp=base_time - timedelta(days=1, seconds=-2),
            all_urls=["https://app.posthog.com/features"],
        )

        flush_persons_and_events()

        # Run the query
        query = ReplayActiveScreensQuery()
        runner = ReplayActiveScreensQueryRunner(query=query, team=self.team)
        response = runner.calculate()

        # Features screen should not appear in results since the session is too short
        features_result = next((r for r in response.results if "features" in r.screen), None)
        assert features_result is None

    def test_query_filters_old_sessions(self):
        # Create a person
        _create_person(
            distinct_ids=["user_old"],
            team=self.team,
        )

        base_time = datetime.now()

        # Create only old sessions (> 7 days ago)
        produce_replay_summary(
            team_id=self.team.pk,
            distinct_id="user_old",
            session_id="old1",
            first_timestamp=base_time - timedelta(days=8),
            last_timestamp=base_time - timedelta(days=8, seconds=-60),
            all_urls=["https://app.posthog.com/old-screen"],
        )

        flush_persons_and_events()

        # Run the query
        query = ReplayActiveScreensQuery()
        runner = ReplayActiveScreensQueryRunner(query=query, team=self.team)
        response = runner.calculate()

        # Old screen should not appear in results since the session is too old
        old_screen_result = next((r for r in response.results if "old-screen" in r.screen), None)
        assert old_screen_result is None

    def test_query_returns_top_10_screens(self):
        # Create a person
        _create_person(
            distinct_ids=["user"],
            team=self.team,
        )

        base_time = datetime.now()

        # Create 15 sessions with different URLs
        for i in range(15):
            produce_replay_summary(
                team_id=self.team.pk,
                distinct_id="user",
                session_id=f"session_{i}",
                first_timestamp=base_time - timedelta(days=1, hours=i),
                last_timestamp=base_time - timedelta(days=1, hours=i, seconds=-60),
                all_urls=[f"https://app.posthog.com/screen{i}"],
            )

        flush_persons_and_events()

        # Run the query
        query = ReplayActiveScreensQuery()
        runner = ReplayActiveScreensQueryRunner(query=query, team=self.team)
        response = runner.calculate()

        # Should return exactly 10 screens (due to LIMIT 10)
        assert len(response.results) <= 10

    @snapshot_clickhouse_queries
    def test_query_performance(self):
        # Create some test data
        _create_person(
            distinct_ids=["perf_user"],
            team=self.team,
        )

        base_time = datetime.now()
        produce_replay_summary(
            team_id=self.team.pk,
            distinct_id="perf_user",
            session_id=str(uuid4()),
            first_timestamp=base_time - timedelta(days=1),
            last_timestamp=base_time - timedelta(days=1, seconds=-60),
            all_urls=["https://app.posthog.com/performance"],
        )
        flush_persons_and_events()

        # Run the query and capture queries
        query = ReplayActiveScreensQuery()
        runner = ReplayActiveScreensQueryRunner(query=query, team=self.team)
        response = runner.calculate()

        assert response.results is not None
