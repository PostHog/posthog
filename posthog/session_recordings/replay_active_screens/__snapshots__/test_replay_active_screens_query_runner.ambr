# serializer version: 1
# name: TestReplayActiveScreensQueryRunner.test_query_performance
  '''
  SELECT cutQueryString(cutFragment(url)) AS screen,
         count(DISTINCT session_id) AS count
  FROM
    (SELECT session_replay_events.session_id AS session_id,
            arrayJoin(any(session_replay_events.all_urls)) AS url
     FROM session_replay_events
     WHERE and(equals(session_replay_events.team_id, 99999), greaterOrEquals(toTimeZone(session_replay_events.min_first_timestamp, 'UTC'), minus(toDateTime64('today', 6, 'UTC'), toIntervalDay(7))), lessOrEquals(toTimeZone(session_replay_events.min_first_timestamp, 'UTC'), toDateTime64('today', 6, 'UTC')))
     GROUP BY session_replay_events.session_id
     HAVING greater(dateDiff('second', min(toTimeZone(session_replay_events.min_first_timestamp, 'UTC')), max(toTimeZone(session_replay_events.max_last_timestamp, 'UTC'))), 5))
  GROUP BY screen
  ORDER BY count DESC
  LIMIT 10 SETTINGS readonly=2,
                    max_execution_time=60,
                    allow_experimental_object_type=1,
                    format_csv_allow_double_quotes=0,
                    max_ast_elements=4000000,
                    max_expanded_ast_elements=4000000,
                    max_bytes_before_external_group_by=0,
                    transform_null_in=1,
                    optimize_min_equality_disjunction_chain_length=4294967295,
                    allow_experimental_join_condition=1
  '''
# ---
